cip/1.accepted/CIP2015-06-24-call-procedures.adoc

```diff
@@ -0,0 +1,319 @@
+
+This CIP proposes adding procedures to Cypher in order to address these issues.
+
+== Proposal
+
+This proposal consists of two parts, introducing procedures conceptually, and describing how to call them in a system that supports Cypher.
+
+=== Procedures
+
+A procedure has a name and takes a fixed number of named and typed arguments in a fixed sequence.
+A procedure produces a stream of result records and may potentially cause a side effect.
+
+Each record produced contains values for the same fixed number of named and typed result fields in the same fixed sequence such that each field value in a yielded record matches its field's type.
+
+Each result field may optionally be marked as *deprecated*.
+
+The name of a procedure, its argument names, and if applicable any result field names follow the same rules as other symbolic names in Cypher (like variables, labels, relationship types, and property keys).
+
+
+==== Procedure signatures
+
+The arguments of a procedure, their types, and their order are called its *argument signature*.
+
+The result fields of returned records, their types, their order, and their deprecation status are called its *result signature*.
+
+Together, the argument signature and the result signature form the (full) *signature* of a procedure.
+
+According to this definition, a procedure may take zero arguments and/or may produce records with zero result fields.
+
+==== Procedure signature notation
+
+This CIP suggests a standard notation for printing procedure names together with their signatures.
+This notation may be used in system monitoring, documentation and error reporting, and perhaps more importantly is to be used in future CIPs.
+It is based on the syntax for type annotations as specified by the Cypher type system:
+
+- Procedures returning a stream of records: `<procName>(arguments) :: (results)`
+- Arguments and results are written as comma-separated lists of `<name> :: <type>`
+- Deprecated results are prefixed with the keyword `DEPRECATED` (i.e. written as `DEPRECATED <name> :: <type>`)
+
+An actual implementation may choose to support a notion of default value for procedure arguments.
+For such a case, this CIP proposes that arguments with default values are written as `<name> = <value> :: <type>`.
+
+==== Procedure implementations
+
+For this proposal, it is assumed that the underlying graph database or graph processing system provides a set of concrete procedure implementations that are understood in terms of their full signatures.
+
+Procedure implementations may be written in different programming languages and provided with APIs for accessing the underlying graph database or graph processing system as appropriate.
+The details of procedure implementation and provisioning are outside the scope of this proposal.
+
```

Comment:
How about adding a sub-section here on procedure naming conventions?
In particular about what name spaces should be used!